### PR TITLE
Set/Update the role visibility for a comment

### PR DESCRIPTION
**Pull request by benjidea**
Set/Update the role visibility for a comment

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description.
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests.
- [x] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description.
- [ ] Reviewed the code.
- [ ] Verified that the appropriate tests have been written or valid explanation given.
- [ ] If applicable, tested by installing this plugin on the Jenkins instance.

### CODE REVIEW

File: src/main/java/org/thoughtslive/jenkins/plugins/jira/service/JiraService.java [hunk 1]
```diff
@@ -187,6 +188,15 @@ public JiraService(final Site jiraSite) {
     }
   }
 
+  public ResponseData<Object> addComment(final String issueIdorKey, final String comment, final String roleVisibility) {
```
Comment (nrayapati) [inline src/main/java/org/thoughtslive/jenkins/plugins/jira/service/JiraService.java]: These method signatures would change to take just object instead: 

Add Comment:
```groovy
  public ResponseData<Object> addComment(final String issueIdorKey, final Object input) {
    try {
      return parseResponse(jiraEndPoints
          .addComment(issueIdorKey, input).execute());
    } catch (Exception e) {
      return buildErrorResponse(e);
    }
  }

```

Update Comment: 
```groovy
  public ResponseData<Object> updateComment(final String issueIdorKey, final String commentId,
      final Object input) {
    try {
      return parseResponse(
          jiraEndPoints.updateComment(issueIdorKey, commentId, input).execute());
    } catch (Exception e) {
      return buildErrorResponse(e);
    }
  }
```

File: hugo/content/changelog.md [hunk 2]
```diff
@@ -6,6 +6,7 @@ lastmodifierdisplayname = "Naresh Rayapati"
 +++
```
Comment (nrayapati) [inline hugo/content/changelog.md]: And also please update the docs at https://github.com/jenkinsci/jira-steps-plugin/tree/master/hugo/content/steps/comment folder appropriately. Thank you.

File: src/test/java/org/thoughtslive/jenkins/plugins/jira/steps/AddCommentStepTest.java [hunk 3]
```diff
@@ -71,6 +71,8 @@ public void setup() throws IOException, InterruptedException {
     doNothing().when(printStreamMock).println();
```
Comment (nrayapati) [inline src/test/java/org/thoughtslive/jenkins/plugins/jira/steps/AddCommentStepTest.java]: These test cases would accordingly to the logic change to the actual steps.

**Comment by benjidea** [comment]
Thanks for your review and great idea!
I've pushed the new implementation.

File: hugo/content/steps/comment/jira_add_comment.md [hunk 4]
```diff
@@ -14,10 +14,11 @@ This step adds a comment on a particular issue.
 #### Input
 
 * **idOrKey** - Issue id or key.
-* **comment** - comment, supports jira wiki formatting.
+* **input** - comment, supports jira wiki formatting.
 * **site** - Optional, default: `JIRA_SITE` environment variable.
 * **failOnError** - Optional. default: `true`.
-* **auditLog** - Optional. default: `true`. Append a panel to the comment with the build url and build user name.
+* _Deprecated:_ **comment** - comment, supports jira wiki formatting.
+* _Deprecated:_ **auditLog** - Optional. default: `true`. Append a panel to the comment with the build url and build user name.
```
Comment (nrayapati) [inline hugo/content/steps/comment/jira_add_comment.md]: Lets not deprecate the `auditLog`.

File: src/main/java/org/thoughtslive/jenkins/plugins/jira/steps/AddCommentStep.java [hunk 5]
```diff
@@ -25,15 +27,25 @@
...
+  private Object input;
+  
+  @Deprecated
   @DataBoundConstructor
   public AddCommentStep(final String idOrKey, final String comment) {
     this.idOrKey = idOrKey;
     this.comment = comment;
   }
 
+  @DataBoundSetter
```
Comment (nrayapati) [inline src/main/java/org/thoughtslive/jenkins/plugins/jira/steps/AddCommentStep.java]: Add `@DataBoundSetter` to the input field, that should be fine, we don't need this setter. Same this applies for Edit step too. 

```java
@Getter 
@DataBoundSetter
private Object input
```

File: src/main/java/org/thoughtslive/jenkins/plugins/jira/steps/AddCommentStep.java [hunk 6]
```diff
@@ -88,14 +106,21 @@ protected Execution(final AddCommentStep step, final StepContext context)
 
       if (response == null) {
         final String idOrKey = Util.fixEmpty(step.getIdOrKey());
-        final String comment = Util.fixEmpty(step.getComment());
+
+        if (step.getComment() != null && step.getInput() != null) {
+          errorMessage = "Use comment or either input.";
```
Comment (nrayapati) [inline src/main/java/org/thoughtslive/jenkins/plugins/jira/steps/AddCommentStep.java]: Use either comment or input.

File: src/main/java/org/thoughtslive/jenkins/plugins/jira/steps/AddCommentStep.java [hunk 7]
```diff
@@ -88,14 +106,21 @@ protected Execution(final AddCommentStep step, final StepContext context)
...
+        }
+
+        if (step.getComment() == null && step.getInput() == null) {
+          errorMessage = "You need to set at least comment or input.";
+        }
 
         if (idOrKey == null) {
```
Comment (nrayapati) [inline src/main/java/org/thoughtslive/jenkins/plugins/jira/steps/AddCommentStep.java]: Lets add couple of util methods in `InputBuilder` class to verify `body` is not null or empty and also to add the metadata (is to maintain the passivity with `auditLog`)

```java
static getField(input, fieldName) {
    return input."$fieldName"
}

static setField(input, fieldName, fieldvalue) {
    input."$fieldName" = fieldvalue
}
```

Get the body (comment) field and also set the metadata by:

```java
String message = InputBuilder.getField(step.input, "body") == null ? InputBuilder.getField(step.input, "body").toString(): "";
InputBuilder.setField(step.input, "body", addMeta(message));
```

File: hugo/content/steps/comment/jira_add_comment.md (hunk 4 above)
Comment (benjidea) [reply hugo/content/steps/comment/jira_add_comment.md]: Fixed.

**Comment by nrayapati** [comment]
I will release this over the weekend once the local testing is successful. Thank you again for this change @benjidea.

**Pull request merged**